src/app-layout/visual-refresh/context.tsx

```diff
@@ -487,6 +490,38 @@ export const AppLayoutInternalsProvider = React.forwardRef(
     const mainElement = useRef<HTMLDivElement>(null);
     const [mainOffsetLeft, setMainOffsetLeft] = useState(0);
 
+    const updateDynamicOverlapHeight = useCallback(
+      (height: number) => {
+        /**
+         * The disableContentHeaderOverlap property is absolute and will always disable the overlap
+         * if it is set to true. If there is no contentHeader then the overlap should be disabled
+         * unless there is a dynamicOverlapHeight. The dynamicOverlapHeight property is set by a
+         * component in the content slot that needs to manually control the overlap height.
+         */
+        const isOverlapSet = !!props.contentHeader || height > 0;
+        const isOverlapDisabled = props.disableContentHeaderOverlap || !isOverlapSet;
+        setIsDynamicOverlapSet(isOverlapSet);
+        setIsDynamicOverlapDisabled(isOverlapDisabled);
+
+        /**
+         * React 18 will trigger a paint before the state is correctly updated
+         * (see https://github.com/facebook/react/issues/24331).
+         * To work around this, we bypass React state updates and imperatively update the custom property on the DOM.
+         * An alternative would be to use `queueMicrotask` and `flushSync` in the ResizeObserver callback,
+         * but that would have some performance impact as it would delay the render.
+         * Using React state for `isDynamicOverlapSet` and `isDynamicOverlapDisabled` is less problematic
+         * because they will rarely change within the lifecycle of an application.
+         */
+        const element = typeof layoutElement === 'function' ? layoutElement(null) : layoutElement?.current;
```

Comment:
This variable is returned by `useContainerQuery` which returns `React.Ref<any>`, which in turn is `RefCallback<any> | RefObject<any> | null`, but its value is set only by the Layout component [here](https://github.com/cloudscape-design/components/blob/main/src/app-layout/visual-refresh/layout.tsx#L87) so we know it's going to be an ObjectRef and not a RefCallback.

If it looks better, we could as well just do nothing if it is a function.

Reply:
Updated.

Reply:
Should `layoutElement` somehow end up being used as a `RefCallback`, it will also cause this `useCallback` to recreate the `updateDynamicOverlapHeight` function on every render.

Anyway, it seems unlikely, so it is also non-blocking for me.